platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,233 @@
+
+import static com.swirlds.platform.util.BootstrapUtils.detectSoftwareUpgrade;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.state.primitives.ProtoBytes;
+import com.hedera.hapi.node.state.roster.Roster;
+import com.hedera.hapi.node.state.roster.RosterState;
+import com.hedera.hapi.node.state.roster.RosterState.Builder;
+import com.hedera.hapi.node.state.roster.RoundRosterPair;
+import com.hedera.hapi.platform.state.PlatformState;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import com.swirlds.common.RosterStateId;
+import com.swirlds.platform.roster.RosterUtils;
+import com.swirlds.platform.roster.RosterValidator;
+import com.swirlds.platform.state.RosterStateAccessor;
+import com.swirlds.platform.state.RosterStateModifier;
+import com.swirlds.platform.state.signed.ReservedSignedState;
+import com.swirlds.platform.system.SoftwareVersion;
+import com.swirlds.state.spi.CommittableWritableStates;
+import com.swirlds.state.spi.WritableKVState;
+import com.swirlds.state.spi.WritableSingletonState;
+import com.swirlds.state.spi.WritableStates;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Read-write interface for accessing rosters states.
+ */
+public class WritableRosterStore implements RosterStateModifier {
+
+    private final WritableStates writableStates;
+    private final RosterStateAccessor rosterStateAccessor;
+    /**
+     * The roster state singleton. This is the state that holds the candidate roster hash and the list of pairs of
+     * active roster hashes and the round number in which those rosters became active.
+     *
+     * @implNote the use of {@link ReadablePlatformStateStore} and {@link WritablePlatformStateStore} to provide access
+     * to the roster states (beyond just the {@link PlatformState}) is deliberate, for convenience.
+     */
+    private final WritableSingletonState<RosterState> rosterState;
+
+    private final WritableKVState<ProtoBytes, Roster> rosterMap;
+
+    /**
+     * Constructs a new {@link WritableRosterStore} instance.
+     *
+     * @param writableStates the readable states
+     */
+    public WritableRosterStore(@NonNull final WritableStates writableStates) {
+        this.writableStates = writableStates;
+        this.rosterStateAccessor = new ReadableRosterStore(writableStates);
+        this.rosterState = writableStates.getSingleton(RosterStateId.ROSTER_STATES_KEY);
+        this.rosterMap = writableStates.get(RosterStateId.ROSTER_KEY);
+    }
+
+    /**
+     * Set the candidate roster if valid and doesn't yet exist in the state.
+     *
+     * @param candidateRoster a candidate roster to set.
+     */
+    @Override
+    public void setCandidateRoster(@NonNull final Roster candidateRoster) {
+        Objects.requireNonNull(candidateRoster);
+        RosterValidator.validate(candidateRoster);
```

Comment:
This is 1 of 2 reasons for the requested changes flag.

Reply:
>The steps to validate and add the Roster to the state will be executed within the same thread that submits the request.
An exception will be thrown if the Roster is invalid. The durability of the submitted Roster is ensured via storing
it in the state.

That's what the proposal says. 

>If roster is not valid, don't set it.

I'm not following the crashing the node bit. Why would this cause a node crash? 
I made it unchecked exception because a `createRoster` is provided which returns a valid roster so if callers used that to create their roster and passes it here, majority of the calls to this path will come with a valid roster. In the few cases they may not, callers will have to handle that themselves but it's important to have a feedback to whether or not the roster is stored. Failing silently by simply not setting it isn't the right approach either I think.

Reply:
My point is we don't know what will happen when someone sets an invalid roster and an unchecked exception is thrown.   If this is thrown in production, then what call stack is disrupted?   I don't think we can assume that the roster was created using your method.   What if the roster is made from scratch or uses a different source of base information?

Reply:
raised an issue https://github.com/hashgraph/hedera-services/issues/15766